### PR TITLE
Normalize PATH on windows to avoid unnecessary full builds

### PR DESCRIPTION
**Pull request by ghost**
I discovered that I was doing unnecessary full builds on windows due to changes in how the PATH environment variable was being presented.

I saw at least 3 different versions of PATH, depending on where I was running ```x.py``` from:
* VS Code
* Windows Terminal
* Windows Terminal launched in Quake mode

In some cases, the PATH would have duplicate entries stripped, and in other cases PATH would have a ```C:\Program Files\WindowsApps\Microsoft.WindowsTerminal_1.12.10733.0_x64__8wekyb3d8bbwe``` path injected into it.

This would cause me to do full builds when switching where I was running ```x.py check``` from, even though I was making no file changes.

This PR normalizes the PATH to always have duplicate entries stripped (while retaining order), and to strip the extra path that Windows Terminal will sometimes inject. With this in place, I can run an ```x.py check``` build from all 3 locations mentioned above without doing an unnecessary full build as I switch between them.

This is a somewhat hacky "developer machine" oriented change... but it's in the ```x.py``` code anyway so that makes sense.

### CODE REVIEW

**Comment by skippy10110** [comment]
cc @jyn514

**Comment by skippy10110** [comment]
i think i found one more version of PATH where it's doing full build, converted to draft while i track down the issue

File: src/bootstrap/bin/main.rs [hunk 1]
```diff
@@ -5,11 +5,40 @@
 //! parent directory, and otherwise documentation can be found throughout the `build`
 //! directory in each respective module.
 
-use std::env;
+use std::{collections::HashSet, env};
 
 use bootstrap::{Build, Config, Subcommand, VERSION};
 
 fn main() {
+    if cfg!(windows) {
```
Comment (jyn514) [inline src/bootstrap/bin/main.rs:13]: I think only the Windows terminal change needs to be gated to Windows. Removing duplicates seems useful in general.

**Comment by skippy10110** [comment]
I'm getting a 4th version of PATH due to these lines in powershell: https://github.com/PowerShell/PowerShell/blob/186d6be47d858460a3d0ba6abd81f779cb633b55/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs#L114-L127

They're always modifying the PATH on startup to stick stuff at the start, which means some of my builds genuinely do have a different PATH. Sometimes with powershell at the start, sometimes in the middle.

I could maybe always ensure that the powershell portion of PATH always ends up at the start, but that's even more hokey...

A documentation fix to say "futz with your PATH in windows to avoid extra builds" might be an alternative if there's no good way to do it.

**Comment by jyn514** [comment]
> I could maybe always ensure that the powershell portion of PATH always ends up at the start, but that's even more hokey...

I don't think x.py should do this. You can add a wrapper script around it if it's a problem for you.

**Comment by jyn514** [comment]
Actually, even removing the Windows Terminal path seems not great now that I think of it, I don't know why Terminal adds that but we're removing it unconditionally even if someone has added it manually.

**Comment by skippy10110** [comment]
I already have an ```x``` function in my profile to mirror the convenience of ```x build``` on linux, so I added a snippet there:
```
function x {
    $env:Path = ($env:Path -split ";" | % { $_.TrimEnd("\/") } | ? { $_ -NotLike "*\WindowsApps\Microsoft.WindowsTerminal_*" } | select -Unique) -join ";"
    $xPy = ,(gi .) + (gi . | Get-ParentDirectories) | % { gcm "$_\x.py" -EA 0 } | select -First 1
    if ($xPy) {
        pushd ($xPy.Source | Split-Path)
        try {
            python.exe $xPy.Source @Args
        } finally {
            popd
        }
    } else {
        Write-HostError "x.py not found. Please run inside of a checkout of `https://github.com/rust-lang/rust`."
    }
}
```

And then to fix the Powershell ordering issue, I change my vs code settings:
```
"rust-analyzer.checkOnSave.overrideCommand": [
    "pwsh.exe"
    "-Command"
    "x",
    "check",
    "--json-output"
],
```
```
"tasks": [
{
    "label": "./x.py check",
    "command": "pwsh.exe -Command x check",
```

So I should (in theory) finally see a consistent PATH this way.

I was ballooning my CPU & disk space usage by having essentially 4 build copies for each PATH variation, so it would be nice to fix this for everyone. Probably just the documentation route is the best bet.

Do you think it's worth keeping this PR up for just the deduplication? The duplicates don't really matter as long as they're consistent.

**Comment by Mark-Simulacrum** [comment]
I would not be inclined to accept this change, at least without a more definitive rationale for why it's causing problems. Where is the dependency on rebuilding when PATH changes coming from? I'm not immediately finding anything that does that outside of bootstrap/build.rs, and that only does so if your rustc is relative. That implies that you can likely fix this for yourself by using `RUSTC=$(rustup which rustc)` or similar.

I think if anything this is another point toward the canonicalization of the rustc binary in bootstrap's build.rs being the wrong approach, and one we should consider replacing (somehow, not super clear how).

**Comment by skippy10110** [comment]
I think writing documentation to save others the headache is my best bet here, I'm going to close. I'm also switching back to WSL, but I would still like to write this all up since it was so disruptive to the workflow and I suspect other Windows devs would be hitting the same thing.

**Pull request closed without merging**